pkg/controller/options.go

```diff
@@ -0,0 +1,96 @@
+package controller
+
+import (
+	"errors"
+
+	clientset "github.com/m3db/m3db-operator/pkg/client/clientset/versioned"
+	informers "github.com/m3db/m3db-operator/pkg/client/informers/externalversions"
+	"github.com/m3db/m3db-operator/pkg/k8sops"
+	"go.uber.org/zap"
+
+	"github.com/m3db/m3db-operator/pkg/m3admin/namespace"
+	"github.com/m3db/m3db-operator/pkg/m3admin/placement"
+
+	kubeinformers "k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+)
+
+// Options configures a new controller.
+type Options struct {
+	logger                     *zap.Logger
+	kclient                    k8sops.K8sops
+	crdClient                  clientset.Interface
+	kubeClient                 kubernetes.Interface
+	kubeInformerFactory        kubeinformers.SharedInformerFactory
+	m3dbClusterInformerFactory informers.SharedInformerFactory
+	namespaceClient            namespace.Client
+	placementClient            placement.Client
+}
+
+// Option configures controller Options.
+type Option func(*Options)
```

Comment:
I am also a fan of this pattern, take a look at a [suggestion from prateek](https://github.com/m3db/m3db-operator/pull/7#discussion_r216458043). Which was implemented within other [packages](https://github.com/m3db/m3db-operator/blob/master/pkg/m3admin/namespace/namespace.go#L70-L91)